---
title: "Differential expression (DESeq2) and GSEA nalysis"
author: "Chao Di, dic@email.chop.edu"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc_depth: 1
    number_sections: yes
  pdf_document:
    toc_depth: '1'
---
## Abstract {-}
This project is for dissecting the transcriptome changes between patients with single gene mutations (STAT3 and STAT1) and HC. The methods used here for differential gene expression analysis is DESeq2.

```{r setup, include=FALSE, echo = FALSE, cache = FALSE, warning=FALSE}
# replace with path where you want the results be
knitr::opts_knit$set(root.dir="/Users/dic/Documents/Sarah Henrickson lab/20210920_Peyton_SHenrickson_RNAseq/")
```

```{r include=FALSE,echo=FALSE,message = FALSE}
# set home dir when test in console 
setwd("/Users/dic/Documents/Sarah Henrickson lab/20210920_Peyton_SHenrickson_RNAseq/")
library(DESeq2)
library(apeglm)
library(ashr)
library(pheatmap)
library(genefilter)
library(ggplot2)
library(RColorBrewer)
library(fdrtool)
library(EnhancedVolcano)
library(fgsea)
library(KEGGREST)
library(EnrichmentBrowser)
library(dplyr)
library(tibble)
library(data.table)
library(org.Hs.eg.db)
```


### Overview of the data {-}
* Load read counts and do data transformation
Supply with raw read counts table "all_readCount.tsv" calculated from STAR/HT-seq, and sample info "sample_contrast.tsv", do value transformation and normalization (variance stabilizing transformations (VST)) in DESeq2.

```{r message=FALSE, warning=FALSE}
cts <- read.table("all_readCount.tsv", header=TRUE, row.names="gene", check.names=FALSE)
colnames(cts) <- sub("_S.+","",colnames(cts))
coldata <- read.table("sample_table.tsv", header=TRUE, row.names="sample",sep="\t", check.names=FALSE)
head(coldata)

# load the counts with no design
dds <- DESeqDataSetFromMatrix(countData=cts, colData=coldata, design=~1)

# genes have at least 10 reads in at least 5 samples
dds <- dds[rowSums(counts(dds) >= 10) >= 5,]

# normalization and pre-processing
dds <- DESeq(dds)

# raw count normalization
norm_counts <- counts(dds, normalized=TRUE) 
# count transformation, log2 scale, either rlog or vst
vsd <- vst(dds, blind=FALSE)
cat("The data values after transformation:\n")
head(assay(vsd)[,1:6])

```

* PCA plot
The PCA plot shows the samples in the 2D plane spanned by their first two principal components. This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects.
```{r message = FALSE, warning=FALSE}
# STAT3GOF
vsd_STAT3GOF <- vsd[, vsd$STAT3GOF_group %in% c("Young", "Adults")]
pcaData <- plotPCA(vsd_STAT3GOF, intgroup=c("disease","STAT3GOF_group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = disease, shape=STAT3GOF_group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()

# STAT3LOF
vsd_STAT3LOF <- vsd[, vsd$STAT3LOF_group %in% c("Young","Older","Oldest")]
pcaData <- plotPCA(vsd_STAT3LOF, intgroup=c("disease","STAT3LOF_group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = disease, shape=STAT3LOF_group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
# STAT1GOF
vsd_STAT1GOF <- vsd[, vsd$STAT1GOF_group %in% c("Young","Older","Adults")]
pcaData <- plotPCA(vsd_STAT1GOF, intgroup=c("disease","STAT1GOF_group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = disease, shape=STAT1GOF_group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()

# STAT1LOF
vsd_STAT1LOF <- vsd[, vsd$STAT1LOF_group %in% c("Youngest")]
pcaData <- plotPCA(vsd_STAT1LOF, intgroup=c("disease","STAT1LOF_group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = disease, shape=STAT1LOF_group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
```

* Heatmap for expression matrix
Heatmap showing the expressions of the most variable genes (top 35 here), optional to do clustering between samples. This gives us the information of which samples and which genes are close to each other.

```{r fig.height=9, fig.width=9, message=FALSE, include=TRUE}
# STAT3GOF
select <- head(order(-rowVars(assay(vsd_STAT3GOF))),50) # most variable genes
pdata <- assay(vsd_STAT3GOF)[select,]
df <- as.data.frame(colData(dds)[,c("disease","STAT3GOF_group")]) %>% filter(STAT3GOF_group!="NA")
pheatmap(
  pdata,
  cluster_rows = TRUE,
  show_rownames = TRUE,
  cluster_cols = TRUE,
  annotation_col = df )

# STAT3LOF
select <- head(order(-rowVars(assay(vsd_STAT3LOF))),50) # most variable genes
pdata <- assay(vsd_STAT3LOF)[select,]
df <- as.data.frame(colData(dds)[,c("disease","STAT3LOF_group")]) %>% filter(STAT3LOF_group!="NA")
pheatmap(
  pdata,
  cluster_rows = TRUE,
  show_rownames = TRUE,
  cluster_cols = TRUE,
  annotation_col = df )

# STAT1GOF
select <- head(order(-rowVars(assay(vsd_STAT1GOF))),50) # most variable genes
pdata <- assay(vsd_STAT1GOF)[select,]
df <- as.data.frame(colData(dds)[,c("disease","STAT1GOF_group")]) %>% filter(STAT1GOF_group!="NA")
pheatmap(
  pdata,
  cluster_rows = TRUE,
  show_rownames = TRUE,
  cluster_cols = TRUE,
  annotation_col = df )


# STAT1LOF
select <- head(order(-rowVars(assay(vsd_STAT1LOF))),50) # most variable genes
pdata <- assay(vsd_STAT1LOF)[select,]
df <- as.data.frame(colData(dds)[,c("disease","STAT1LOF_group")]) %>% filter(STAT1LOF_group!="NA")
pheatmap(
  pdata,
  cluster_rows = TRUE,
  show_rownames = TRUE,
  cluster_cols = TRUE,
  annotation_col = df )

```


### Differential expression analysis by `DESeq2` {-}
Calculation of expression values (log2 scale), transformed by variance stabilizing transformations (VST).
Setup contrast as disease vs HC in different age groups: Young, Adults, Older, Oldest etc (grouping variate for different diseases). Export tables with the standard DESeq2 result format including "baseMean log2FoldChange lfcSE stat pvalue padj" plus the "Log fold change shrinked" normalized readcounts )

* STAT3GOF 
```{r message=FALSE, warning = FALSE}
dds_STAT3GOF <- dds[,dds$STAT3GOF_group %in% c("Young", "Adults")]
dds_STAT3GOF$disease <- factor(dds_STAT3GOF$disease, levels = c("HC","STAT3GOF"))
dds_STAT3GOF$STAT3GOF_group <- factor(dds_STAT3GOF$STAT3GOF_group, levels=c("Young", "Adults"))

dds_STAT3GOF$group <- factor(paste0(dds_STAT3GOF$STAT3GOF_group, dds_STAT3GOF$disease))
design(dds_STAT3GOF) <- ~ group
dds_STAT3GOF <- DESeq(dds_STAT3GOF)
resultsNames(dds_STAT3GOF)
res_STAT3GOF_Young <- results(dds_STAT3GOF, contrast=c("group", "YoungSTAT3GOF","YoungHC"))
res_STAT3GOF_Adults <- results(dds_STAT3GOF, contrast=c("group", "AdultsSTAT3GOF","AdultsHC"))
table(res_STAT3GOF_Young$padj<0.1)
data.frame(res_STAT3GOF_Young[!is.na(res_STAT3GOF_Young$padj) & res_STAT3GOF_Young$padj<0.1,])
res_STAT3GOF_Young[rownames(res_STAT3GOF_Young)=="STAT3",]

table(res_STAT3GOF_Adults$padj<0.1)
data.frame(res_STAT3GOF_Adults[!is.na(res_STAT3GOF_Adults$padj) & res_STAT3GOF_Adults$padj<0.1,])
res_STAT3GOF_Adults[rownames(res_STAT3GOF_Adults)=="STAT3",]

```
* STAT3LOF 
```{r message=FALSE, warning = FALSE}
dds_STAT3LOF <- dds[,dds$STAT3LOF_group %in% c("Young","Older","Oldest")]
dds_STAT3LOF$disease <- factor(dds_STAT3LOF$disease, levels = c("HC","STAT3LOF"))
dds_STAT3LOF$STAT3LOF_group <- factor(dds_STAT3LOF$STAT3LOF_group, levels =c("Young","Older","Oldest"))

dds_STAT3LOF$group <- factor(paste0(dds_STAT3LOF$disease, dds_STAT3LOF$STAT3LOF_group))
design(dds_STAT3LOF) <- ~ group
dds_STAT3LOF <- DESeq(dds_STAT3LOF)
resultsNames(dds_STAT3LOF)
res_STAT3LOF_Young <- results(dds_STAT3LOF, contrast=c("group", "STAT3LOFYoung","HCYoung"))
res_STAT3LOF_Older <- results(dds_STAT3LOF, contrast=c("group", "STAT3LOFOlder","HCOlder"))
res_STAT3LOF_Oldest <- results(dds_STAT3LOF, contrast=c("group", "STAT3LOFOldest","HCOldest"))
table(res_STAT3LOF_Young$padj<0.1)
res_STAT3LOF_Young[!is.na(res_STAT3LOF_Young$padj) & res_STAT3LOF_Young$padj<0.1,]
res_STAT3LOF_Young[rownames(res_STAT3LOF_Young)=="STAT3",]

table(res_STAT3LOF_Older$padj<0.1)
res_STAT3LOF_Older[!is.na(res_STAT3LOF_Older$padj) & res_STAT3LOF_Older$padj<0.1,]
res_STAT3LOF_Older[rownames(res_STAT3LOF_Older)=="STAT3",]

table(res_STAT3LOF_Oldest$padj<0.1)
res_STAT3LOF_Oldest[!is.na(res_STAT3LOF_Oldest$padj) & res_STAT3LOF_Oldest$padj<0.1,]
res_STAT3LOF_Oldest[rownames(res_STAT3LOF_Oldest)=="STAT3",]
```


* STAT1GOF 
```{r message=FALSE, warning = FALSE}
dds_STAT1GOF <- dds[,dds$STAT1GOF_group %in% c("Young","Older","Adults")]
dds_STAT1GOF$disease <- factor(dds_STAT1GOF$disease, levels = c("HC","STAT1GOF"))
dds_STAT1GOF$STAT1GOF_group <- factor(dds_STAT1GOF$STAT1GOF_group, levels =c("Young","Older","Adults"))

dds_STAT1GOF$group <- factor(paste0(dds_STAT1GOF$disease, dds_STAT1GOF$STAT1GOF_group))
design(dds_STAT1GOF) <- ~ group
dds_STAT1GOF <- DESeq(dds_STAT1GOF)
resultsNames(dds_STAT1GOF)
res_STAT1GOF_Young <- results(dds_STAT1GOF, contrast=c("group", "STAT1GOFYoung","HCYoung"))
res_STAT1GOF_Older <- results(dds_STAT1GOF, contrast=c("group", "STAT1GOFOlder","HCOlder"))
res_STAT1GOF_Adults <- results(dds_STAT1GOF, contrast=c("group", "STAT1GOFAdults","HCAdults"))
table(res_STAT1GOF_Young$padj<0.1)
res_STAT1GOF_Young[!is.na(res_STAT1GOF_Young$padj) & res_STAT1GOF_Young$padj<0.1,]
res_STAT1GOF_Young[rownames(res_STAT1GOF_Young)=="STAT1",]

table(res_STAT1GOF_Older$padj<0.1)
res_STAT1GOF_Older[!is.na(res_STAT1GOF_Older$padj) & res_STAT1GOF_Older$padj<0.1,]
res_STAT1GOF_Older[rownames(res_STAT1GOF_Older)=="STAT1",]

table(res_STAT1GOF_Adults$padj<0.1)
res_STAT1GOF_Adults[!is.na(res_STAT1GOF_Adults$padj) & res_STAT1GOF_Adults$padj<0.1,]
res_STAT1GOF_Adults[rownames(res_STAT1GOF_Adults)=="STAT1",]
```

* STAT1LOF 
```{r message=FALSE, warning = FALSE}
dds_STAT1LOF <- dds[,dds$STAT1LOF_group %in% c("Youngest")]
dds_STAT1LOF$disease <- factor(dds_STAT1LOF$disease, levels = c("HC","STAT1LOF"))
design(dds_STAT1LOF) <- ~ disease
dds_STAT1LOF <- DESeq(dds_STAT1LOF)
resultsNames(dds_STAT1LOF)
res_STAT1LOF <- results(dds_STAT1LOF, contrast=c("disease", "STAT1LOF","HC"))
table(res_STAT1LOF$padj<0.1)
res_STAT1LOF[!is.na(res_STAT1LOF$padj) & res_STAT1LOF$padj<0.1,]
res_STAT1LOF[rownames(res_STAT1LOF)=="STAT1",]
```


### Additional P-value inspection and correction {-}
Use fdrtool to correct the overestimated p-value, https://www.huber.embl.de/users/klaus/Teaching/DESeq2Predoc2014.html
(depends on the pvalue distribution, optional)
```{r }
# # STAT3GOF Young
# distribution of raw pvalues
# hist(res_STAT3GOF_Young$pvalue)
# res_STAT3GOF_Young <- res_STAT3GOF_Young[!is.na(res_STAT3GOF_Young$pvalue),]
# res_STAT3GOF_Young <- res_STAT3GOF_Young[!is.na(res_STAT3GOF_Young$padj),]
# res_STAT3GOF_Young <- res_STAT3GOF_Young[,-which(names(res_STAT3GOF_Young)=="padj")]
# FDR.res_STAT3GOF_Young <- fdrtool(res_STAT3GOF_Young$stat, statistic="normal", plot=F)
# res_STAT3GOF_Young[,"padj"]  <- p.adjust(FDR.res_STAT3GOF_Young$pval, method = "BH")
# # distribution of correct pvalues
# hist(FDR.res_STAT3GOF_Young$pval)
# table(res_STAT3GOF_Young$padj<0.1)
## Check the histgram of raw p-values
# # STAT3GOF Adults
# hist(res_STAT3GOF_Adults$pvalue)
# # STAT3LOF Young
# # distribution of raw pvalues
# hist(res_STAT3LOF_Young$pvalue)
# # STAT3LOF Older
# hist(res_STAT3LOF_Older$pvalue)
# # STAT3LOF Oldest
# hist(res_STAT3LOF_Oldest$pvalue)
# # STAT1GOF Young
# hist(res_STAT1GOF_Young$pvalue)
# # STAT1GOF Older
# hist(res_STAT1GOF_Older$pvalue)
# # STAT1GOF Adults
# hist(res_STAT1GOF_Adults$pvalue)
# # STAT1LOF
# hist(res_STAT1LOF$pvalue)
```

```{r message=FALSE, warning = FALSE, fig.height=8, fig.width=7, fig.show="hold"}
for (each in c("STAT3GOF_Young", "STAT3GOF_Adults", "STAT3LOF_Young", "STAT3LOF_Older", "STAT3LOF_Oldest","STAT1GOF_Young", "STAT1GOF_Older", "STAT1GOF_Adults", "STAT1LOF")){
  comp <- paste0(each, "_disease_vs_HC") 
  res = get(paste0("res_",each))
  top10 = res[order(res$padj),][1:10,]
  
  # only label padj<0.1 as significant
keyvals <- ifelse(
     rownames(res) %in% rownames(res[!is.na(res$padj) & res$padj<0.1,]), 'red', 'blue')
names(keyvals)[keyvals == 'red'] <- 'padj<0.1'
names(keyvals)[keyvals == 'blue'] <- 'Non-significant'

  volcano <- EnhancedVolcano(
      res,
      lab = rownames(res),
      selectLab = rownames(top10),
      x = 'log2FoldChange',
      y = 'pvalue',
      ylim = c(0, max(-log10(res$pvalue) + 1)),
      title = comp,
      subtitle = "",
      pCutoff = 0.05,
      FCcutoff = 2,
      pointSize = 2,
      labSize = 4,
      colCustom = keyvals,
      drawConnectors = TRUE,
      widthConnectors = 0.5,
      typeConnectors = 'open',
      colConnectors = 'grey') +
      theme(plot.title = element_text(hjust = 0.5))

  print(volcano)
}
```

### Gene set enrichment analyis {-}
```{r include=FALSE, echo=FALSE, message=FALSE, warning = FALSE}
# create a function for GSEA and plot
gseafunc <- function(gmtfile, pval_cutoff, gene_sets){
    for (each in c("STAT3GOF_Young", "STAT3GOF_Adults", "STAT3LOF_Young", "STAT3LOF_Older", 
                   "STAT3LOF_Oldest","STAT1GOF_Young",     "STAT1GOF_Older", "STAT1GOF_Adults", "STAT1LOF")){
      # print comparison
      comp <- paste0(each, "_disease_vs_HC") 
      print(comp)
      res = get(paste0("res_",each))
      res <- res[order(res$stat,decreasing = T),]
      res$symbol = rownames(res)
      ranked_genes <- res %>% as.data.frame() %>% dplyr::select(symbol, stat) %>% deframe() 
      fgseaRes <- fgsea(pathways = gmtfile, 
                        stats = ranked_genes,
                        minSize=15,
                        maxSize=600,
                        eps = 0)
      
      # select most significant pathways to show in a table plot
      # topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=10), pathway]
      # topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=10), pathway]
      # topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
      # plotGseaTable(gmtfile[topPathways], ranked_genes, fgseaRes, gseaParam = 0.5)
    
      # select independent pathways, and filter by pval_cutoff
      collapsedPathways <- collapsePathways(fgseaRes[order(pval)][padj < pval_cutoff], gmtfile, ranked_genes)
      fgseaRes_filtered <- fgseaRes[pathway %in% collapsedPathways$mainPathways][order(-NES), ]
      print(head(fgseaRes_filtered))
      
      # plot in fgsea table style
      # mainPathways <- fgseaRes[pathway %in% collapsedPathways$mainPathways][order(-NES), pathway]
      # plotGseaTable(gmtfile[mainPathways], ranked_genes, fgseaRes, gseaParam = 0.5)
        
      # output the table
      fwrite(fgseaRes_filtered, file=paste0("enriched_genesets_", gene_sets,"_",comp,".tsv"), sep="\t", sep2=c("",",",""))
      
      p <- ggplot(fgseaRes_filtered, aes(reorder(pathway, NES), NES, color=padj)) +
           geom_point(aes(size=size)) +
           scale_color_gradient(low="red", high="blue") +
           #scale_fill_manual(values = c("Down-regulated" = "dodgerblue","Up-regulated" = "firebrick") ) +
           coord_flip() +
           labs(x="Pathway", y="Normalized Enrichment Score", title= comp, color="adjust.p", size="Count") + 
           theme_minimal()
      print(p)
    }
}
```

* Hallmark gene sets
```{r include=TRUE, echo=FALSE, message=FALSE, warning = FALSE, fig.show="hold"}
# load genesets in .gmt format
gmtfile = fgsea::gmtPathways("./geneSets/h.all.v7.2.symbols.gmt")
gseafunc(gmtfile=gmtfile, pval_cutoff=0.25,gene_sets="GSEA-Hallmarks")
```

* C2.Curated gene sets -KEGG
```{r echo=FALSE, message=FALSE, warning = FALSE, fig.show="hold"}
# load genesets in .gmt format
gmtfile = fgsea::gmtPathways("./geneSets/c2.cp.kegg.v7.2.symbols.gmt")
gseafunc(gmtfile=gmtfile, pval_cutoff=0.25, gene_sets="GSEA-C2_KEGG")
```


<!-- ## C7.Immunological signature gene sets (running slow) --> 
<!-- ```{r fig.height=20, fig.show="hold", fig.width=12, message=FALSE, warning=FALSE} -->
<!-- # load genesets in .gmt format -->
<!-- gmtfile = fgsea::gmtPathways("./geneSets/c7.all.v7.2.symbols.gmt") -->
<!-- gseafunc(gmtfile=gmtfile, pval_cutoff=0.25, gene_sets="GSEA-Immunological signature") -->
<!-- ``` -->


* Exhaustion gene set (Wherry lab)
```{r message=FALSE, warning = FALSE, fig.height=2, fig.width=4, fig.show="hold"}
# load genesets in .gmt format
exhaustion_up = read.table("./geneSets/ESG_UP_ATAC.grp.txt",header=F)
exhaustion_down = read.table("./geneSets/ESG_DN_ATAC.grp.txt",header=F)
gmtfile = list(ESG=c(as.vector(exhaustion_up[[1]]), as.vector(exhaustion_down[[1]])))

for (each in c("STAT3GOF_Young", "STAT3GOF_Adults", "STAT3LOF_Young", "STAT3LOF_Older", 
                   "STAT3LOF_Oldest","STAT1GOF_Young", "STAT1GOF_Older", "STAT1GOF_Adults", "STAT1LOF")){
      # print comparison
      comp <- paste0(each, "_disease_vs_HC") 
      print(comp)
      res = get(paste0("res_",each))
      res <- res[order(res$stat,decreasing = T),]
      res$symbol = rownames(res)
      ranked_genes <- res %>% as.data.frame() %>% dplyr::select(symbol, stat) %>% deframe() 
      ranked_genes <- na.omit(ranked_genes)
      fgseaRes <- fgsea(pathways = gmtfile, 
                        stats = ranked_genes,
                        minSize=15,
                        maxSize=600,
                        eps = 0)
     
      print(fgseaRes)
      
      # output the table
      fwrite(fgseaRes, file=paste0("enriched_genesets_Exhaustion_genesets","_",comp,".tsv"), sep="\t", sep2=c("",",",""))
      
      # enrichment plot
      print(plotEnrichment(gmtfile[["ESG"]], ranked_genes) + labs(title=paste0(comp,'\npadj = ', formatC(fgseaRes$padj, format="e", digits = 2))))
      
}

```